### ISSUE TITLE
Performance degradation 0.3.0 -> 1.0.0 -> 1.0.2

### ISSUE THREAD

**Issue by richie** [exAspArk/batch-loader#2]
I'm using batch-loader in conjunction with the graphql gem, and I've been experimenting with updating the gem version from 0.3.0 to 1.0.2. In doing so, I noticed a performance decrease that I thought I would report.

Our javascript client does an expensive graphQL call after login to download much of the static data it needs to display throughout a single page app. A couple of the resources requested during this expensive 'getEnums' call rely on batch-loader. On my development machine, if I perform this query with 0.3.0, it anecdotally takes circa 500ms on average. Here are 3 logins
```
INFO [16/Sep/2017:09:28:05 +0100]: GraphQL [getCurrentUser] completed in 12 ms
INFO [16/Sep/2017:09:28:05 +0100]: GraphQL [getEnums] completed in 627 ms
INFO [16/Sep/2017:09:28:38 +0100]: GraphQL [getCurrentUser] completed in 12 ms
INFO [16/Sep/2017:09:28:38 +0100]: GraphQL [getEnums] completed in 649 ms
INFO [16/Sep/2017:09:28:59 +0100]: GraphQL [getCurrentUser] completed in 8 ms
INFO [16/Sep/2017:09:28:59 +0100]: GraphQL [getEnums] completed in 491 ms
```
If I upgrade to 1.0.0, it takes approx 800ms to return. 
```
INFO [16/Sep/2017:09:33:23 +0100]: GraphQL [getCurrentUser] completed in 12 ms
INFO [16/Sep/2017:09:33:24 +0100]: GraphQL [getEnums] completed in 819 ms
INFO [16/Sep/2017:09:33:40 +0100]: GraphQL [getCurrentUser] completed in 11 ms
INFO [16/Sep/2017:09:33:41 +0100]: GraphQL [getEnums] completed in 839 ms
INFO [16/Sep/2017:09:33:58 +0100]: GraphQL [getCurrentUser] completed in 8 ms
INFO [16/Sep/2017:09:33:59 +0100]: GraphQL [getEnums] completed in 732 ms
```
Same is true for 1.0.1 (actually seems slightly faster in these test runs):
```
INFO [16/Sep/2017:09:36:20 +0100]: GraphQL [getCurrentUser] completed in 10 ms
INFO [16/Sep/2017:09:36:21 +0100]: GraphQL [getEnums] completed in 852 ms
INFO [16/Sep/2017:09:36:32 +0100]: GraphQL [getCurrentUser] completed in 6 ms
INFO [16/Sep/2017:09:36:33 +0100]: GraphQL [getEnums] completed in 733 ms
INFO [16/Sep/2017:09:36:47 +0100]: GraphQL [getCurrentUser] completed in 6 ms
INFO [16/Sep/2017:09:36:47 +0100]: GraphQL [getEnums] completed in 703 ms
```
If I upgrade to 1.0.2 however, this shoots up to 2800ms or so.
```
INFO [16/Sep/2017:09:39:32 +0100]: GraphQL [getCurrentUser] completed in 11 ms
INFO [16/Sep/2017:09:39:35 +0100]: GraphQL [getEnums] completed in 2811 ms
INFO [16/Sep/2017:09:39:49 +0100]: GraphQL [getCurrentUser] completed in 7 ms
INFO [16/Sep/2017:09:39:51 +0100]: GraphQL [getEnums] completed in 2665 ms
INFO [16/Sep/2017:09:40:06 +0100]: GraphQL [getCurrentUser] completed in 12 ms
INFO [16/Sep/2017:09:40:08 +0100]: GraphQL [getEnums] completed in 2751 ms
```
The only variation between the tests is updating the Gemfile and altering batch_loader.load to batch_loader.call.

In terms of data retrieved using batch-loader it's currently restricted to small portion of the data returned in the call. We have Clouds containing SubClouds, containing ProductFamilies and batch-loader is used to fetch these in fewer queries.

graphql/types/cloud.rb

```ruby
  field :subClouds, -> { types[SubCloudType] }, 'The sub-clouds available for this cloud',
        resolve: lambda { |cloud, _args, _ctx|
          BatchLoader.for(cloud.id).batch do |cloud_ids, batch_loader|
            sub_clouds = SubCloud.in(cloud_id: cloud_ids).entries
            cloud_ids.each do |cloud_id|
              sub = sub_clouds.select { |s| s.cloud_id == cloud_id }
              batch_loader.call(cloud_id, sub)
            end
          end
        }
```

graphql/types/subcloud.rb

```ruby
  field :productFamilies, -> { types[ProductFamilyType] }, 'The product families available for this sub-cloud',
        resolve: lambda { |sub_cloud, _args, _ctx|
          BatchLoader.for(sub_cloud.id).batch do |sub_cloud_ids, batch_loader|
            families = ProductFamily.in(subCloud_id: sub_cloud_ids).entries
            sub_cloud_ids.each do |sub_cloud_id|
              family = families.select { |f| f.subCloud_id == sub_cloud_id }
              batch_loader.call(sub_cloud_id, family)
            end
          end
        }
```

I'll see if I can log the lambda calls and database queries to see what the underlying differences are.

**Comment by richie** [comment]
I must be using it wrong in the 1.x versions. 

I put logging in the 3 GraphQL types that use the batch_loader. In 0.3.0, I see 3 underlying calls to the DB with the appropriate batch of IDs for each type. In 1.0.x, I see 114 separate calls, each having just a singular ID.

**Comment by exAspArk** [comment]
Hey, @richie!

Thank you so much for reporting the issue and providing the detailed description!

I think these things can help solve the slow performance:

* Use `BatchLoader::GraphQL` in GraphQL schema

There were some breaking changes in the `v1.0.0`:

> Now syncing is done implicitly when you call any method on the lazy object

Automatic syncing is useful if you use `BatchLoader` without GraphQL ( GraphQL syncs the lazy objects automatically). 

But I haven't highlighted it in the changelog that people should also use `use BatchLoader::GraphQL` in the [GraphQL schema definition](https://github.com/exAspArk/batch-loader#graphql-example) instead of `lazy_resolve BatchLoader, :sync`. I hope that if you apply these changes to your code, it'll batch these N+1 queries.

* You can optimize you `batch` block to avoid iterating through an array for each element. 

So, instead of:

```ruby
families = ProductFamily.in(subCloud_id: sub_cloud_ids).entries
sub_cloud_ids.each do |sub_cloud_id|
  family = families.select { |f| f.subCloud_id == sub_cloud_id } # is it O(NM), which is similar to in some cases O(N^2)?
  loader.call(sub_cloud_id, family)
end
```

Iterate through `families` just once (the same can be applied for "clouds"):

```ruby
families = ProductFamily.in(subCloud_id: sub_cloud_ids).entries
families.each do |family|
  loader.call(family.subCloud_id, family) # since v1.0.0 it's better to call the variable like "loader" which is a lambda, not BatchLoader instance
end
```

But it, of course, depends on a number of elements in the arrays. If it's small then it won't have much difference. 

Please let me know if it helps to solve the performance issue.

**Comment by exAspArk** [comment]
Oh, I just realized that you use:

```ruby
family = families.select { |f| f.subCloud_id == sub_cloud_id }
loader.call(sub_cloud_id, family)
```

So, `select` returns an array, not a single `family` as I thought :)

In this case, I can suggest using building a hash (1 iteration) and then access it (`O(1)`). For example:

```ruby
families = ProductFamily.in(subCloud_id: sub_cloud_ids).entries
families_by_cloud_id = Hash.new { |h, k| h[k] = [] } # key is an array by default
families.each { |family| families_by_cloud_id[family.id] << family } # e.g. { '1' => [family1, family2] }

sub_cloud_ids.each do |sub_cloud_id|
  loader.call(sub_cloud_id, families_by_cloud_id[sub_cloud_id])
end
```

**Comment by richie** [comment]
Thanks, the schema change helped immensely. I can see that the calls are again batched and the execution times are comparable:

```
INFO [16/Sep/2017:21:40:48 +0100]: GraphQL [getEnums] completed in 705 ms
INFO [16/Sep/2017:21:41:11 +0100]: GraphQL [getEnums] completed in 701 ms
INFO [16/Sep/2017:21:41:34 +0100]: GraphQL [getEnums] completed in 637 ms
```

Thanks for the quick response and this fantastic software!

**Comment by richie** [comment]
Above was with 1.0.1.

With 1.0.2, I can see that the calls are resolved in a batch but the response times are high:
```
INFO [16/Sep/2017:21:48:13 +0100]: GraphQL [getEnums] completed in 2838 ms
```
Will see if I can see anything obvious in the database queries

**Comment by exAspArk** [comment]
Hey @richie!

I profiled `1.0.2` and identified a performance degradation because of Ruby `Forwardable` with `def_delegators` which I decided to use in this version. Up to 30x slower in some micro-benchmarks. If you have time, could you please try:

```ruby
# Gemfile
gem 'batch-loader', github: 'exAspArk/batch-loader', branch: 'perf'
```

I expect that with the `perf` branch you'll see the same 600-700ms. If so, I'll release these changes.

**Comment by richie** [comment]
@exAspArk thanks so much for looking into this. That branch makes a huge difference:
```
INFO [17/Sep/2017:10:27:35 +0100]: GraphQL [getEnums] completed in 694 ms
INFO [17/Sep/2017:10:28:02 +0100]: GraphQL [getEnums] completed in 675 ms
INFO [17/Sep/2017:10:28:16 +0100]: GraphQL [getEnums] completed in 682 ms
```

**Comment by exAspArk** [comment]
@richie awesome, thank you!

I was also experimenting with BatchLoader + GraphQL performance specifically. In some cases, I was able to improve performance up to 3x times. I wonder if with the `better-perf` branch you'll be able to get 500-600ms?

```ruby
# Gemfile
gem 'batch-loader', github: 'exAspArk/batch-loader', branch: 'better-perf'
```

I'll try to test this branch with the real application in production environment and release it next week.

**Comment by richie** [comment]
The results are in:
```
INFO [17/Sep/2017:20:54:46 +0100]: GraphQL [getEnums] completed in 626 ms
INFO [17/Sep/2017:20:55:34 +0100]: GraphQL [getEnums] completed in 476 ms
INFO [17/Sep/2017:20:55:56 +0100]: GraphQL [getEnums] completed in 638 ms
INFO [17/Sep/2017:20:56:27 +0100]: GraphQL [getEnums] completed in 490 ms
```
Amazing! 🥇

**Comment by exAspArk** [comment]
Hi @richie!

I released the performance fixes in `v1.0.3`.
 
Thank you so much for finding time to report the issue and making `batch-loader` better!  🙌
